### DISCUSSION TITLE
Failed build succeeds on rerun using cache

### DISCUSSION THREAD

**Discussion by dnillovna** [webpack/webpack#18478] [Q&A]
# Bug report

**What is the current behavior?**

My build runs: `npm run build`
On the first run, it fails. A file in dependency has a typescript error, causing the build to fail. 
When I try to rerun it, it succeeds. 
When I turn off the cache, it fails every time.

Example package to reproduce: https://github.com/dnillovna/webpack-cache-fail.git

**What is the expected behavior?**

Failed builds should not create a cache that can be used for future builds. When cache is enabled, the build with errors should fail every time instead of only on the first attempt.

**Other relevant information:**

webpack version:

> webpack: latest => 5.91.0
> webpack-cli: latest => 5.1.4

Node.js version:

>   Node: 21.6.1 - /opt/homebrew/opt/node/bin/node
>   npm: 10.2.4 - /opt/homebrew/opt/node/bin/npm

Operating System:

>   OS: macOS 13.6.2
>   CPU: (8) arm64 Apple M1 Pro
>   Memory: 100.47 MB / 32.00 GB

Additional tools:
  None
  
I got the information using `npm run version-info`, which executes `webpack --version`.

I found a similar issue for the other webpack cache plugin: https://github.com/mzgoddard/hard-source-webpack-plugin/issues/507

**Comment by alexander-akait** [comment]
Please open an issue in ts-loader, they don't emit a warning/error using webpack API, so webpack thinks everything is okay and store a module in cache, we can't fix it here

I want to make it discussion, because we can't fix it here

/cc @johnnyreilly Why ts-loader doesn't use webpack API to emit warnings and errors?

**Reply by alexander-akait** [reply]
I see the problem, related https://github.com/TypeStrong/ts-loader/issues/1332, because we emit errors in the hook, it is too late and our code already cached, we need to migrate from hook on pure loader API, so it will solve type caching and the problems with error emiting

**Reply by alexander-akait** [reply]
I can try to rewrite ts-loader logic, but it requires time and maybe major release, because some develpers can use hooks to make some internal things

/cc @johnnyreilly if it is fine for you we can plan it
